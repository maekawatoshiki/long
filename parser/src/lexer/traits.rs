use anyhow::Result;
use ast::token::{kind::TokenKind, Token};
use std::path::PathBuf;

pub trait LexerLike {
    /// Returns the next token.
    fn next(&mut self) -> Result<Option<Token>>;

    /// Peeks the next token.
    fn peek(&mut self) -> Result<Option<Token>>;

    /// Ungets the given `token`.
    fn unget(&mut self, token: Token);

    /// Skips the next token only if it matches the given kind.
    fn skip(&mut self, kind: TokenKind) -> bool;

    /// Returns the file path the lexer currently reads.
    fn filepath(&self) -> Option<&PathBuf>;
}

pub struct TokenStream {
    tokens: Vec<Token>,
    buf: Vec<Token>,
    cur: usize,
}

impl TokenStream {
    /// Creates a new `TokenStream`.
    pub fn new(tokens: Vec<Token>) -> Self {
        Self {
            tokens,
            buf: Vec::new(),
            cur: 0,
        }
    }
}

impl LexerLike for TokenStream {
    fn next(&mut self) -> Result<Option<Token>> {
        if let Some(tok) = self.buf.pop() {
            return Ok(Some(tok));
        }
        self.cur += 1;
        Ok(self.tokens.get(self.cur - 1).cloned())
    }

    fn peek(&mut self) -> Result<Option<Token>> {
        if let Some(tok) = self.buf.last().cloned() {
            return Ok(Some(tok));
        }
        Ok(self.tokens.get(self.cur).cloned())
    }

    fn unget(&mut self, token: Token) {
        self.buf.push(token);
    }

    fn skip(&mut self, kind: TokenKind) -> bool {
        if let Ok(Some(tok)) = self.next() {
            if tok.kind() == &kind {
                return true;
            }
            self.unget(tok);
        }
        false
    }

    fn filepath(&self) -> Option<&PathBuf> {
        None
    }
}

impl From<Vec<Token>> for TokenStream {
    fn from(tokens: Vec<Token>) -> Self {
        Self::new(tokens)
    }
}

#[test]
fn vec_tokens() {
    use ast::token::*;
    use sourceloc::SourceLoc;
    let tokens = vec![
        Token::new(
            kind::TokenKind::Ident("foo".to_string()),
            SourceLoc::new(0, 1),
        ),
        Token::new(
            kind::TokenKind::Int(kind::IntKind::Int(123)),
            SourceLoc::new(0, 5),
        ),
    ];
    let ll: &mut dyn LexerLike = &mut TokenStream::from(tokens);
    assert!(ll.next().unwrap().is_some());
    assert!(ll.next().unwrap().is_some());
    assert!(ll.next().unwrap().is_none());
}

#[test]
fn vec_tokens_skip() {
    use ast::token::*;
    use sourceloc::SourceLoc;
    let tokens = vec![
        Token::new(kind::SymbolKind::Add.into(), SourceLoc::new(0, 1)),
        Token::new(kind::IntKind::Int(123).into(), SourceLoc::new(0, 5)),
    ];
    let ll: &mut dyn LexerLike = &mut TokenStream::from(tokens);
    assert!(ll.skip(kind::SymbolKind::Add.into()));
    assert!(!ll.skip(kind::SymbolKind::Add.into()));
    assert!(ll.skip(kind::IntKind::Int(123).into()));
    assert!(!ll.skip(kind::IntKind::Int(123).into()));
}
