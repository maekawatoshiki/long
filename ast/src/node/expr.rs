use super::lit::Literal;
use super::Located;
use crate::token::kind::IntKind;

/// An expression kind.
#[derive(Debug, Clone)]
pub enum Expr {
    /// A literal expression.
    Literal(Literal),

    /// A unary expression.
    Unary(UnaryOp, Box<Located<Expr>>),

    /// A binary expression.
    Binary(BinOp, Box<Located<Expr>>, Box<Located<Expr>>),

    /// A ternary expression.
    Ternary(Box<Located<Expr>>, Box<Located<Expr>>, Box<Located<Expr>>), // cond, then, else.

    /// An Assign expression.
    Assign(AssignOp, Box<Located<Expr>>, Box<Located<Expr>>),
}

/// A unary operator kind.
#[derive(Debug, Clone, PartialEq)]
pub enum UnaryOp {
    Not,
}

/// A binary operator kind.
#[derive(Debug, Clone, PartialEq)]
pub enum BinOp {
    Comma,
    LogicalOr,
    LogicalAnd,
    And,
    Eq,
    Ne,
    Or,
    Gt,
    Ge,
    Lt,
    Le,
    Shl,
    Shr,
    Add,
    Sub,
    Mul,
    Div,
    Rem,
}
/// An assignment operator kind.
/// `=  *=  /=  %=   +=  -=  >>=  <<=  &=  ^=  |=`
/// <https://timsong-cpp.github.io/cppwp/n3337/expr.ass#1>
#[derive(Debug, Clone, PartialEq)]
pub enum AssignOp {
    Add,
    Sub,
    Mul,
    Div,
    Rem,
    Shl,
    Shr,
    And,
    Or,
    Xor,
    None,
}

impl Located<Expr> {
    // TODO: Add tests for this.
    /// Evaluates the constant expression.
    pub fn eval_constexpr(&self) -> Option<i64> {
        match self.inner_ref() {
            Expr::Unary(UnaryOp::Not, ref val) => Some((val.eval_constexpr() == Some(0)) as i64),
            Expr::Literal(Literal::Int(IntKind::Int(i))) => Some(*i as i64),
            Expr::Literal(_) => todo!(),
            Expr::Binary(BinOp::Comma, _, ref rhs) => rhs.eval_constexpr(),
            Expr::Binary(BinOp::LogicalOr, ref lhs, ref rhs) => {
                if lhs.eval_constexpr() != Some(0) {
                    Some(1)
                } else {
                    Some((rhs.eval_constexpr() != Some(0)) as i64)
                }
            }
            Expr::Binary(BinOp::LogicalAnd, ref lhs, ref rhs) => {
                if lhs.eval_constexpr() == Some(0) {
                    Some(0)
                } else {
                    Some((rhs.eval_constexpr() != Some(0)) as i64)
                }
            }
            Expr::Binary(ref op, ref lhs, ref rhs) => {
                let lhs = lhs.eval_constexpr()?;
                let rhs = rhs.eval_constexpr()?;
                match op {
                    BinOp::And => Some(lhs & rhs),
                    BinOp::Or => Some(lhs | rhs),
                    BinOp::Shl => Some(lhs << rhs),
                    BinOp::Shr => Some(lhs >> rhs),
                    BinOp::Eq => Some((lhs == rhs) as i64),
                    BinOp::Ne => Some((lhs != rhs) as i64),
                    BinOp::Gt => Some((lhs > rhs) as i64),
                    BinOp::Ge => Some((lhs >= rhs) as i64),
                    BinOp::Lt => Some((lhs < rhs) as i64),
                    BinOp::Le => Some((lhs <= rhs) as i64),
                    BinOp::Add => Some(lhs + rhs),
                    BinOp::Sub => Some(lhs - rhs),
                    BinOp::Mul => Some(lhs * rhs),
                    BinOp::Div => Some(lhs / rhs),
                    BinOp::Rem => Some(lhs % rhs),
                    BinOp::Comma => unreachable!(),
                    BinOp::LogicalAnd => unreachable!(),
                    BinOp::LogicalOr => unreachable!(),
                }
            }
            Expr::Ternary(ref cond, ref thn, ref els) => {
                if cond.eval_constexpr() != Some(0) {
                    thn.eval_constexpr()
                } else {
                    els.eval_constexpr()
                }
            }
            Expr::Assign(AssignOp::None, _, ref rhs) => rhs.eval_constexpr(),
            Expr::Assign(_, _, _) => todo!(),
        }
    }
}
