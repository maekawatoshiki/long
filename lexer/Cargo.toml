[package]
name = "long-lexer"
version = "0.1.0"
edition = "2021"

[dependencies]
long-sourceloc = { path = "../sourceloc" }
anyhow = "= 1.0.45"

[dev-dependencies]
insta = "= 1.7.1"
