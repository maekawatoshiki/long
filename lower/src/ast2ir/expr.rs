use super::LowerCtx;
use anyhow::Result;
use long_ast::{
    node::{
        expr::{AssignOp, Expr as AstExpr},
        lit::Literal as AstLiteral,
        Located,
    },
    token::kind::IntKind,
};
use long_ir::{
    expr::{Expr as IrExpr, Literal as IrLiteral},
    name::Name,
};

pub fn lower_expr<'a>(ctx: &mut LowerCtx<'a>, expr: &AstExpr) -> Result<&'a IrExpr<'a>> {
    match expr {
        AstExpr::Literal(AstLiteral::Int(IntKind::Int(i))) => Ok(ctx
            .ir_ctx
            .expr_arena
            .alloc(IrExpr::Literal(IrLiteral::Int(IntKind::Int(*i))))),
        AstExpr::Literal(AstLiteral::Ident(ident)) => {
            let name = ctx.ir_ctx.name_arena.alloc(Name::local(ident));
            let var = ctx.envs.lookup_local(name).expect("TODO");
            Ok(ctx
                .ir_ctx
                .expr_arena
                .alloc(IrExpr::Literal(IrLiteral::Local(var))))
        }
        AstExpr::Assign(AssignOp::None, lhs, rhs) => {
            let lhs = Located::new(lower_expr(ctx, &lhs.inner)?, lhs.loc());
            let rhs = Located::new(lower_expr(ctx, &rhs.inner)?, lhs.loc());
            Ok(ctx
                .ir_ctx
                .expr_arena
                .alloc(IrExpr::Assign(AssignOp::None, lhs, rhs)))
        }
        _ => todo!(),
    }
}
