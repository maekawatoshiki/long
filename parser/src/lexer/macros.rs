use ast::token::Token;
use std::collections::HashMap;

/// Defined macros.
#[derive(Debug, Clone)]
pub struct Macros(HashMap<String, Macro>);

/// A macro body.
#[derive(Debug, Clone)]
pub enum Macro {
    /// Object-like macro.
    Obj(Vec<Token>),

    /// Function-like macro.
    Func(Vec<FuncMacroToken>),
}

/// A function-like macro token.
/// `Param(leading space, n)` represents a `n`th parameter.
/// `Vararg(leading space, n)` represents the `n`th parameter is `...`.
// TODO: Param and Vararg should not have leading space as bool.
#[derive(Debug, Clone)]
pub enum FuncMacroToken {
    Token(Token),
    Param(bool, usize),
    Vararg(bool, usize),
}

impl Macros {
    /// Creates a new `Macros`.
    pub fn new() -> Self {
        Self(HashMap::new())
    }

    /// Adds a new object-like macro.
    pub fn add_obj_macro(&mut self, name: impl Into<String>, body: Vec<Token>) {
        self.0.insert(name.into(), Macro::Obj(body));
    }

    /// Adds a new function-like macro.
    pub fn add_func_macro(&mut self, name: impl Into<String>, body: Vec<FuncMacroToken>) {
        self.0.insert(name.into(), Macro::Func(body));
    }

    /// If `name` is a defined macro name, returns the corresponding macro body.
    pub fn find(&self, name: impl AsRef<str>) -> Option<&Macro> {
        self.0.get(name.as_ref())
    }

    pub fn remove(&mut self, name: impl AsRef<str>) {
        self.0.remove(name.as_ref());
    }
}
