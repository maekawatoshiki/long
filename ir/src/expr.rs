use long_ast::node::{
    expr::{AssignOp, BinOp, UnaryOp},
    Located,
};

/// An expression kind.
pub enum Expr<'a> {
    // /// A literal expression.
    // Literal(Literal),
    /// A unary expression.
    Unary(UnaryOp, Located<&'a Expr<'a>>),

    /// A binary expression.
    Binary(BinOp, Located<&'a Expr<'a>>, Located<&'a Expr<'a>>),

    /// A ternary expression.
    Ternary(
        Located<&'a Expr<'a>>,
        Located<&'a Expr<'a>>,
        Located<&'a Expr<'a>>,
    ), // cond, then, else.

    /// An Assign expression.
    Assign(AssignOp, Located<&'a Expr<'a>>, Located<&'a Expr<'a>>),
}
