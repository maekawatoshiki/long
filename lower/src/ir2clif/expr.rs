use anyhow::Result;
use cranelift::{
    frontend::FunctionBuilder,
    prelude::{InstBuilder, Value},
};
use cranelift_codegen::ir::types as clif_ty;
use long_ast::token::kind::IntKind;
use long_ir::expr::{Expr, Literal};

pub fn lower_expr(builder: &mut FunctionBuilder, expr: &Expr) -> Result<Value> {
    match expr {
        Expr::Literal(Literal::Int(IntKind::Int(i))) => {
            Ok(builder.ins().iconst(clif_ty::I32, *i as i64))
        }
        _ => todo!(),
    }
}
