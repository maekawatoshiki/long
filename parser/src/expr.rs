use crate::lexer::{traits::LexerLike, Error};
use crate::Parser;
use anyhow::Result;
use long_ast::node::expr::UnaryOp;
use long_ast::{
    node::{
        expr::{BinOp, Expr},
        lit::Literal,
        Located,
    },
    token::kind::{SymbolKind, TokenKind},
};

macro_rules! def_binop {
    ($name:ident < $child:ident, $($sym:ident -> $op:ident),+) => {
        fn $name(&mut self) -> Result<Located<Expr>> {
            let mut lhs = self.$child()?;
            let loc = lhs.loc();
            loop {
                $(
                    if self.lexer.skip(SymbolKind::$sym.into()) {
                        let rhs = self.$child()?;
                        lhs = Located::new(
                            Expr::Binary(BinOp::$op, Box::new(lhs), Box::new(rhs)),
                            loc,
                        );
                        continue;
                    }
                )+;
                break;
            }
            Ok(lhs)
        }
    };
}

impl<'a, L: LexerLike> Parser<'a, L> {
    /// Parses a comma expression.
    pub(crate) fn parse_comma(&mut self) -> Result<Located<Expr>> {
        let mut lhs = self.parse_assign()?;
        let loc = lhs.loc();
        while self.lexer.skip(SymbolKind::Comma.into()) {
            let rhs = self.parse_assign()?;
            lhs = Located::new(
                Expr::Binary(BinOp::Comma, Box::new(lhs), Box::new(rhs)),
                loc,
            );
        }
        Ok(lhs)
    }

    /// Parses an assignment expression.
    pub(crate) fn parse_assign(&mut self) -> Result<Located<Expr>> {
        let mut lhs = self.parse_logor()?;
        let loc = lhs.loc();

        if self.lexer.skip(SymbolKind::Question.into()) {
            return self.parse_ternary(lhs);
        }

        loop {
            // TODO: Support more assignment operators.
            if self.lexer.skip(SymbolKind::Assign.into()) {
                let rhs = self.parse_assign()?;
                lhs = Located::new(Expr::Assign(Box::new(lhs), Box::new(rhs)), loc);
                continue;
            }
            break;
        }

        Ok(lhs)
    }

    /// Parses a ternary expression.
    fn parse_ternary(&mut self, cond: Located<Expr>) -> Result<Located<Expr>> {
        let loc = cond.loc();
        let thn = self.parse_assign()?;
        self.expect(SymbolKind::Colon)?;
        let els = self.parse_assign()?;
        Ok(Located::new(
            Expr::Ternary(Box::new(cond), Box::new(thn), Box::new(els)),
            loc,
        ))
    }

    def_binop!(parse_logor < parse_logand, LOr -> LogicalOr);
    def_binop!(parse_logand < parse_or, LAnd -> LogicalAnd);
    def_binop!(parse_or < parse_and, Or -> Or);
    def_binop!(parse_and < parse_eq_ne, And -> And);
    def_binop!(parse_eq_ne < parse_lt_le_gt_ge, Eq -> Eq, Ne -> Ne);
    def_binop!(parse_lt_le_gt_ge < parse_shl_shr, Lt -> Lt, Le -> Le, Gt -> Gt, Ge -> Ge);
    def_binop!(parse_shl_shr < parse_add_sub, Shl -> Shl, Shr -> Shr);
    def_binop!(parse_add_sub < parse_mul_div_rem, Add -> Add, Sub -> Sub);
    def_binop!(parse_mul_div_rem < parse_unary, Asterisk -> Mul, Div -> Div, Mod -> Rem);

    /// Parses a unary expression.
    fn parse_unary(&mut self) -> Result<Located<Expr>> {
        let tok = self.lexer.peek()?.ok_or(Error::UnexpectedEof)?;
        match tok.kind() {
            TokenKind::Symbol(SymbolKind::Not) => {
                self.lexer.next()?;
                return Ok(Located::new(
                    Expr::Unary(UnaryOp::Not, Box::new(self.parse_unary()?)),
                    *tok.loc(),
                ));
            }
            _ => self.parse_primary(),
        }
    }

    /// Parses a literal or parenthetical expression.
    pub(crate) fn parse_primary(&mut self) -> Result<Located<Expr>> {
        // TODO: Support other than int literals.
        let tok = self.lexer.next()?.ok_or(Error::UnexpectedEof)?;
        match tok.kind() {
            TokenKind::Int(i) => Ok(Located::new(Expr::Literal(Literal::Int(*i)), *tok.loc())),
            TokenKind::Symbol(SymbolKind::OpeningParen) => {
                let expr = self.parse_expr()?;
                self.expect(SymbolKind::ClosingParen)?;
                Ok(expr)
            }
            e => Err(Error::Message(format!("Unimplemented: {:?}", e), *tok.loc()).into()),
        }
    }
}

#[test]
fn parse_comma() {
    use crate::lexer::Lexer;
    let node = Parser::new(&mut Lexer::new("1, 2, 3")).parse_expr();
    insta::assert_debug_snapshot!(node);
}

#[test]
fn parse_ternary() {
    use crate::lexer::Lexer;
    let node = Parser::new(&mut Lexer::new("1 ? 2 : 3")).parse_expr();
    insta::assert_debug_snapshot!(node);
}

#[test]
fn parse_logor_logand() {
    use crate::lexer::Lexer;
    let node = Parser::new(&mut Lexer::new("1 && 2 || 3 && 4")).parse_expr();
    insta::assert_debug_snapshot!(node);
}

#[test]
fn parse_paren() {
    use crate::lexer::Lexer;
    let node = Parser::new(&mut Lexer::new("0 && (1 || 2)")).parse_expr();
    insta::assert_debug_snapshot!(node);
}

#[test]
fn parse_unary() {
    use crate::lexer::Lexer;
    let node = Parser::new(&mut Lexer::new("!0")).parse_expr();
    insta::assert_debug_snapshot!(node);
}

#[test]
fn parse_unary_logand() {
    use crate::lexer::Lexer;
    let node = Parser::new(&mut Lexer::new("!0 && !1")).parse_expr();
    insta::assert_debug_snapshot!(node);
}

#[test]
fn parse_lt_le_gt_ge() {
    use crate::lexer::Lexer;
    let node = Parser::new(&mut Lexer::new("0 < 1 && 0 <= 1 && 0 > 1 && 0 >= 1")).parse_expr();
    insta::assert_debug_snapshot!(node);
}

#[test]
fn parse_add_sub() {
    use crate::lexer::Lexer;
    let node = Parser::new(&mut Lexer::new("0 + 1 - 2")).parse_expr();
    insta::assert_debug_snapshot!(node);
}

#[test]
fn parse_shl_shr() {
    use crate::lexer::Lexer;
    let node = Parser::new(&mut Lexer::new("0 << 1 >> 2")).parse_expr();
    insta::assert_debug_snapshot!(node);
}

#[test]
fn parse_and_or() {
    use crate::lexer::Lexer;
    let node = Parser::new(&mut Lexer::new("0 & 1 | 2")).parse_expr();
    insta::assert_debug_snapshot!(node);
}

#[test]
fn parse_eq_ne() {
    use crate::lexer::Lexer;
    let node = Parser::new(&mut Lexer::new("0 == 0 && 0 != 1")).parse_expr();
    insta::assert_debug_snapshot!(node);
}

#[test]
fn parse_mul_div_rem() {
    use crate::lexer::Lexer;
    let node = Parser::new(&mut Lexer::new("0 * 1 / 2 % 3")).parse_expr();
    insta::assert_debug_snapshot!(node);
}

#[test]
fn parse_assign() {
    use crate::lexer::Lexer;
    let node = Parser::new(&mut Lexer::new("0 = 1 = 2")).parse_expr();
    insta::assert_debug_snapshot!(node);
}
