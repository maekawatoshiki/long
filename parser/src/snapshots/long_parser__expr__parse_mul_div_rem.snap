---
source: parser/src/expr.rs
expression: node

---
Ok(
    Expr {
        kind: Binary(
            Rem,
            Expr {
                kind: Binary(
                    Div,
                    Expr {
                        kind: Binary(
                            Mul,
                            Expr {
                                kind: Literal(
                                    Int(
                                        Int(
                                            0,
                                        ),
                                    ),
                                ),
                                loc: None:1:1,
                            },
                            Expr {
                                kind: Literal(
                                    Int(
                                        Int(
                                            1,
                                        ),
                                    ),
                                ),
                                loc: None:1:5,
                            },
                        ),
                        loc: None:1:1,
                    },
                    Expr {
                        kind: Literal(
                            Int(
                                Int(
                                    2,
                                ),
                            ),
                        ),
                        loc: None:1:9,
                    },
                ),
                loc: None:1:1,
            },
            Expr {
                kind: Literal(
                    Int(
                        Int(
                            3,
                        ),
                    ),
                ),
                loc: None:1:13,
            },
        ),
        loc: None:1:1,
    },
)
